actors/miner/src/lib.rs

```diff
@@ -1152,43 +1150,40 @@ impl Actor {
                 continue;
             }
 
-            let deal_spaces = match activate_deals_and_claim_allocations(
-                rt,
-                update.deals.clone(),
-                sector_info.expiration,
-                sector_info.sector_number,
-            )? {
-                Some(deal_spaces) => deal_spaces,
-                None => {
-                    info!(
-                        "failed to activate deals on sector {}, skipping from replica update set",
-                        update.sector_number
-                    );
-                    continue;
-                }
-            };
+            update_sector_infos.push(UpdateAndSectorInfo { update, sector_info });
+        }
 
-            let expiration = sector_info.expiration;
-            let seal_proof = sector_info.seal_proof;
-            validated_updates.push(UpdateAndSectorInfo { update, sector_info, deal_spaces });
+        let activation_infos: Vec<DealsActivationInfo> = update_sector_infos
+            .iter()
+            .map(|usi| DealsActivationInfo {
+                deal_ids: usi.update.deals.clone(),
+                sector_number: usi.update.sector_number,
+                sector_expiry: usi.sector_info.expiration,
+            })
+            .collect();
+        let deals_spaces = batch_activate_deals_and_claim_allocations(rt, &activation_infos)?;
 
-            sectors_deals.push(ext::market::SectorDeals {
-                sector_type: seal_proof,
-                deal_ids: update.deals.clone(),
-                sector_expiry: expiration,
-            });
-            sectors_data_spec.push(ext::market::SectorDataSpec {
-                sector_type: seal_proof,
-                deal_ids: update.deals.clone(),
-            });
-        }
+        // associate the successfully activated sectors with the ReplicaUpdateInner and SectorOnChainInfo
+        let validated_updates: Vec<(UpdateAndSectorInfo, ext::market::DealSpaces)> = deals_spaces
+            .into_iter()
+            .zip(update_sector_infos)
+            .filter_map(|(maybe_deal_space, info)| maybe_deal_space.map(|deal| (info, deal)))
+            .collect();
 
         if validated_updates.is_empty() {
             return Err(actor_error!(illegal_argument, "no valid updates"));
         }
 
-        // Errors past this point cause the prove_replica_updates call to fail (no more skipping sectors)
+        let sectors_deals: Vec<ext::market::SectorDeals> = validated_updates
+            .iter()
+            .map(|(usi, _)| ext::market::SectorDeals {
+                deal_ids: usi.update.deals.clone(),
+                sector_expiry: usi.sector_info.expiration,
+                sector_type: usi.sector_info.seal_proof,
+            })
+            .collect();
 
+        // Errors past this point cause the prove_replica_updates call to fail (no more skipping sectors)
         let deal_data = request_deal_data(rt, &sectors_deals)?;
```

Comment:
I've opened https://github.com/filecoin-project/builtin-actors/issues/1308 to follow up on removing this (after the other optimisations you are pursuing).